tensorflow/compiler/tf2tensorrt/convert/convert_nodes.cc

```diff
@@ -1454,38 +1454,169 @@ Status Converter::RenameAndMarkOutputTensors(
-                           int32_t* selection) noexcept override {
+                           int32_t nbChoices, int32_t* selection) override {
+
+    std::unordered_set<int64_t> cherrypicked_tactics;
+    std::unordered_set<int64_t> rejected_tactics;
+
     // TensorRT always provides more than zero number of algorithms
     // in selectAlgorithms.
     assert(nbChoices > 0);
 
-    // making sure that the requested TRT algorithm ID doesn't go above the
-    // max value accepted.
-    selection[0] = std::min(algorithm_id_, nbChoices);
-    return 1;
+    // Disable all disallowed tactics
+    for (auto i = 0; i < nbChoices; i++) {
+      if (algoChoices[i] == nullptr) {
+        continue;
+      }
+
+      int64_t tacticId = algoChoices[i]->getAlgorithmVariant().getTactic();
+
+      if (disallowed_tactics.find(tacticId) != disallowed_tactics.end()) {
+        VLOG(1) << "Rejecting a disallowed tactic: " << tacticId
+                << " for node " << algoContext.getName();
+        rejected_tactics.insert(i);
+        continue;
+      }
+
+#if !IS_TRT_VERSION_GE(8, 0, 0, 0)
+      int64_t implementation =
+          algoChoices[i]->getAlgorithmVariant().getImplementation();
+      nvinfer1::TensorFormat format =
+          algoChoices[i]->getAlgorithmIOInfo(0).getTensorFormat();
+      assert(tacticId != static_cast<int64_t>(Tactic::kINVALID_TACTIC));
+
+      // Reject shuffle node when input format is 32-wide channel vectorized
+      // row major FP32 format
+      if (implementation == static_cast<int64_t>(LayerImpl::kSHUFFLE) &&
+          format == nvinfer1::TensorFormat::kCHW32) {
+        VLOG(1) << "Rejecting a disallowed tactic: " << tacticId
+                << " for node " << algoContext.getName()
+                << " with implementation " << implementation
+                << " and input format " << static_cast<int32_t>(format);
+        rejected_tactics.insert(i);
+        continue;
+      }
+#endif  // !IS_TRT_VERSION_GE(8, 0, 0, 0)
+
+      cherrypicked_tactics.insert(i);
+    }
+
+    if (! allowed_tactics.empty()) {
+      assert(allowed_tactics.size() == 1);
+
+      // Clearing previously selected tactics.
+      cherrypicked_tactics.clear();
+
+      int64_t algorithm_id = *allowed_tactics.begin(0);
+      VLOG(1) << "Forcing TRT algorithm selection to: ID=" << algorithm_id;
+      // Making sure that the requested TRT algorithm ID doesn't go above
+      // the max value accepted.
+      int64_t forced_tactic = std::min(algorithm_id,
+                                      static_cast<int64_t>(nbChoices));
+
+      while (true) {
+        if (forced_tactic > static_cast<int64_t>(nbChoices)) {
+          VLOG(1) << "Impossible to force TensorRT tactic selection. All "
+                  << "attempts have been rejected. Please select a different "
+                  << "value for `TF_TRT_FIXED_ALGORITHM_ID`. Received: "
+                  << algorithm_id << ".";
+          assert(false);
+        }
```

Comment:
The assignment contradicts with the if condition here?

Reply:
My bad. Typo.
Not necessary anymore with the new code